webapp/src/main/java/com/box/l10n/mojito/service/image/S3ImageService.java

```diff
@@ -0,0 +1,69 @@
+package com.box.l10n.mojito.service.image;
+
+import com.box.l10n.mojito.entity.Image;
+import com.box.l10n.mojito.service.blobstorage.s3.S3BlobStorage;
+import org.slf4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import java.util.Optional;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Service to upload and retrieve images from S3.
+ * <p>
+ * If a requested image cannot be found in S3 an attempt will be made to retrieve from the Mojito DB, if the image is
+ * found in the DB then an async task will be executed to upload the image to S3.
+ * <p>
+ * Configured {@link S3BlobStorage} and {@link com.amazonaws.services.s3.AmazonS3} client instances are required to
+ * upload and retrieve images from S3.
+ *
+ * @author maallen
+ */
+@Component
+@ConditionalOnProperty(value = "l10n.blob-storage.s3.images.enabled", havingValue = "true")
+public class S3ImageService implements ImageService {
+
+    /**
+     * logger
+     */
+    static Logger logger = getLogger(S3ImageService.class);
+
+    @Autowired
+    S3BlobStorage s3BlobStorage;
+
+    @Autowired
+    S3UploadImageAsyncTask s3UploadImageAsyncTask;
+
+    @Autowired
+    ImageRepository imageRepository;
```

Comment:
should this be delegated to `DatabaseImageService`?

Reply:
The DatabaseImageService bean won't be present in the container if the S3ImageService is active which is why I used the ImageRepository bean. 

If the DatabaseImageService is active we'd have two active ImageService beans in the container and would then need to qualify which bean type to use in ImageWS & ThirdPartyService in different circumstances, I think keeping the specific implementation hidden from the higher-level beans is cleaner.

Reply:
Can we have something more similar to `BlobStorageConfiguration`, instead of enable/disable it can have different values ` @ConditionalOnProperty(value = "l10n.blob-storage.type", havingValue = "s3")`. 

So 3 types `database`, `s3`, `s3Fallback` and then some configration like that do deal with multi instance:

```
@ConditionalOnProperty(value = "l10n.blob-storage.type", havingValue = "s3Fallback")`. 
static class S3FallbackImageServiceConfiguration {
    @Bean @Qualifier("database") 
    database()

    @Bean @Qualifier("s3") 
    s3()

     @Bean @Primary
    s3Fallback(@Qualifier("database) db, @Qualifier("s3")
}

@ConditionalOnProperty(value = "l10n.blob-storage.type", havingValue = "s3")`. 
static class S3ImageServiceConfiguration {
    @Bean 
    s3()
}

@ConditionalOnProperty(value = "l10n.blob-storage.type", havingValue = "database")`. 
static class DatabaseImageServiceConfiguration {
    @Bean 
    database()
}
```

Reply:
Sure I can add something more flexible